online-prediction/predict/main.py

```diff
@@ -4,65 +4,57 @@
-    logging.info("Model Loaded!")
+@app.get('/health')
+async def health():
+    return Response(status_code=200)
+
+
+@app.post('/init')
+def init(request: InitializeRequest):
+    logging.info("Online Prediction: Init Request:{}".format(request))
+
     try:
-        # If the model artifact is of type `dict`
-        if isinstance(model, dict):
-            categorical_cols = model["cat_columns"] if "cat_columns" in model else []
-            numerical_cols = [x for x in df.columns if x not in categorical_cols]
-            # Transforming the input data-frame using encoder & normalizer from the experiment artifact
-            if ("encoder" in model) or ("normalizer" in model):
-                x_encoded = model["encoder"].transform(
-                    df[categorical_cols]).toarray() if "encoder" in model and categorical_cols else []
-                x_normalized = model["normalizer"].transform(df[numerical_cols]) if "normalizer" in model else df[
-                    numerical_cols].values
-                if np.any(x_encoded) and np.any(x_normalized):
-                    x_transformed = np.concatenate((x_encoded, x_normalized), axis=1)
-                else:
-                    x_transformed = x_encoded if np.any(x_encoded) else x_normalized
-            else:
-                x_transformed = df.values
-            predictions = model["model"].predict(x_transformed)
-        else:
-            # If the model object is an instance of model itself
-            predictions = model.predict(instances)
+        load_model(request.api_endpoint, request.token, request.project_id, request.properties.experiment_name,
+                   request.properties.run_id)
+    except ValueError as e:
+        logging.error(traceback.format_exc())
+        raise HTTPException(status_code=400, detail=str(e))
     except Exception as e:
-        raise Exception("Error occurred while making predictions, Please check the model. Message: {}".format(e))
-    return {'payload': predictions.tolist()}
+        logging.error(traceback.format_exc())
+        raise HTTPException(status_code=500, detail=str(e))
 
+    return {
+        "payload": "model loaded successfully"
+    }
+
+
+@app.post('/invoke')
+async def run(request: InvokeRequest):
+    logging.info("Online Prediction: Invoke Request:{}".format(request))
 
-async def load_model(client, experiment_name, run_id):
     try:
-        experiment = client.experiment(experiment_name)
-        run = experiment.get_run(run_id) if run_id else experiment.last_run()
-        return run.get_artifact('model')
+        predictions = await process(request)
+    except ValueError as e:
+        raise HTTPException(status_code=400, detail=str(e))
     except Exception as e:
-        logging.error("Error: Failed to load model: {}".format(e))
+        logging.error(traceback.format_exc())
+        raise HTTPException(status_code=500, detail=str(e))
```

Comment:
500 is the generic error that Cortex will return if something fails (ie.  the container won't start).  Can you change this to a 400 series ie.. 412 Precondition failed?

Reply:
Agreed. 500 should be reserved for Cortex / Server errors.